pkg/preflight/validate.go

```diff
@@ -0,0 +1,401 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to get existing temp PVC %s: %w", tmpPVCSpec.Name, err)
+		}
+	}
+
+	// consume pvc to determine any access mode errors
+	pvcConsumerPodSpec := buildTmpPVCConsumerPod(tmpPVC.Name, tmpPVC.Namespace, tmpPodImage)
+	pvcConsumerPod, err := client.CoreV1().Pods(tmpPVC.Namespace).Create(ctx, pvcConsumerPodSpec, metav1.CreateOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create validation pod %s: %w", pvcConsumerPodSpec.Name, err)
+	}
+
+	// cleanup pvc and pod at the end
+	defer func() {
+		// pod must be deleted first then the pvc
+		if err = deletePVConsumerPod(client, pvcConsumerPod); err != nil {
+			l.Printf("failed to cleanup pv consumer pod %s: %s", pvcConsumerPod.Name, err)
+		}
+
+		if err = deleteTmpPVC(l, client, tmpPVC); err != nil {
+			l.Printf("failed to cleanup tmp claim: %s", err)
+		}
+	}()
+
+	podReadyTimeoutEnd := time.Now().Add(timeout)
+	for {
+		gotPod, err := client.CoreV1().Pods(pvcConsumerPodSpec.Namespace).Get(ctx, pvcConsumerPodSpec.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get pv consumer pod %s: %w", gotPod.Name, err)
+		}
+
+		switch {
+		case k8spodutils.IsPodReady(gotPod):
+			return nil, nil
+		default:
+			time.Sleep(time.Second)
+		}
+
+		if time.Now().After(podReadyTimeoutEnd) {
+			// The volume consumer pod never went into running phase which means it's probably an error
+			// with provisioning the volume.
+			// A pod in Pending phase means the API Server has created the resource and stored it in etcd,
+			// but the pod has not been scheduled yet, nor have container images been pulled from the registry.
+			if gotPod.Status.Phase == corev1.PodPending {
+				// check pvc status and get error
+				pvcPendingError, err := getPVCError(client, tmpPVC)
+				if err != nil {
+					return nil, fmt.Errorf("failed to get pvc failure: %w", err)
+				}
+				return pvcPendingError, nil
+			}
+			// pod failed for other reason(s)
+			return nil, fmt.Errorf("unexpected status for pod %s: %s", gotPod.Name, gotPod.Status.Phase)
+		}
+	}
+}
+
+// deleteTmpPVC deletes the provided pvc from the default namespace and waits until the
+// backing pv dissapear as well (this is mandatory so we don't leave any orphan pv as this would
+// cause pvmigrate to fail). this function has a timeout of 5 minutes, after that an error is
+// returned.
+func deleteTmpPVC(l *log.Logger, client k8sclient.Interface, pvc *corev1.PersistentVolumeClaim) error {
+	// Cleanup should use background context so as not to fail if context has already been canceled
+	ctx := context.Background()
+
+	pvs, err := client.CoreV1().PersistentVolumes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list persistent volumes: %w", err)
+	}
+
+	pvsByPVCName := map[string]corev1.PersistentVolume{}
+	for _, pv := range pvs.Items {
+		if pv.Spec.ClaimRef == nil {
+			continue
+		}
+		pvsByPVCName[pv.Spec.ClaimRef.Name] = pv
```

Comment:
I can't stop thinking that this function (`deleteTmpPVC`) will timeout if we have two PVCs with the same `name` in two different `namespaces`. Not sure even if this is a valid scenario but I got this itch that does not stop.

Reply:
This indeed seems to be the case. It looks like you inspired this code on the `kurl` implementation, I found the same problem there: https://github.com/replicatedhq/kURL/pull/3837

Reply:
Thanks @ricardomaraschini. I'll fix this.

Reply:
The way the migrate code deals with this is by using a map[string]map[string]corev1.PersistentVolume{}

The fix here will (afaict) just skip one of the two if there's a name collision?

Reply:
@laverya Correct. In my latest update, I'm filtering the list of PVs by the namespace where the soon-to-be-deleted `pvc` resides. This approach will avoid the name collision for PVCs in other namespaces.

Reply:
As long as it works if there's to-be-deleted PVCs with the same name in multiple namespaces at once 😄